core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -2574,22 +2567,34 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
         {
             // determine super type
             Type superType = UNKNOWN;
+
+            ListMultimap<Type, Expression> typeExpressions = LinkedListMultimap.create();
             for (Expression expression : expressions) {
-                Optional<Type> newSuperType = typeCoercion.getCommonSuperType(superType, process(expression, context));
+                typeExpressions.put(process(expression, context), expression);
+            }
+
+            // We need an explicit copy to avoid ConcurrentModificationException
+            Set<Type> types = ImmutableSet.copyOf(typeExpressions.keySet());
+
+            for (Type type : types) {
+                Optional<Type> newSuperType = typeCoercion.getCommonSuperType(superType, type);
                 if (newSuperType.isEmpty()) {
-                    throw semanticException(TYPE_MISMATCH, expression, message, superType);
+                    throw semanticException(TYPE_MISMATCH, typeExpressions.get(type).get(0), message, superType);
                 }
                 superType = newSuperType.get();
             }
 
             // verify all expressions can be coerced to the superType
-            for (Expression expression : expressions) {
-                Type type = process(expression, context);
+            for (Type type : types) {
+                List<Expression> coercionCandidates = typeExpressions.removeAll(type);
+
                 if (!type.equals(superType)) {
                     if (!typeCoercion.canCoerce(type, superType)) {
-                        throw semanticException(TYPE_MISMATCH, expression, message, superType);
+                        throw semanticException(TYPE_MISMATCH, typeExpressions.get(type).get(0), message, superType);
```

Comment:
this `typeExpressions.get(type).get(0)` won't work as you have just removed all elements for type